### DISCUSSION TITLE
Is it possible to use SingleStreamProjection<T> with snapshots?

### DISCUSSION THREAD

**Discussion by federico-paolillo** [JasperFx/marten#2641] [Q&A] [answered]
I want to enable snapshots on a document for which I have already a class extending `SingleStreamProjection<T>`. This class is named `SensorProjection` and is defined defined as follows:

```csharp
public sealed class SensorProjection : SingleStreamProjection<Sensor>
```

  I wanted to reuse `SensorProjection` so I tried to register it as a snapshot as follows:

```csharp
Projections.Snapshot<SensorProjection>(SnapshotLifecycle.Inline);
```

This fails. Marten complains that `SensorProjection` does not have suitable `Apply()` methods because Marten expects the `Apply()` method to receive the same `T` configured with `Projections.Snapshot<T>`. It seems that[ Marten instantiates a `SingleStreamProjection<T>` when configuring snapshots.](https://github.com/JasperFx/marten/blob/999d371d2c84b397aa1e62cc544bfc55ebbc39fa/src/Marten/Events/Projections/ProjectionOptions.cs#L200)

Is there a way to use `SingleStreamProjection<T>` with snapshots or only document that directly mutates themselves are currently supported ?

**Comment by mysticmind** [comment]
You have to understand the concepts of usage clearly, outlining it below:
- Projections are read models created by projecting/shaping events optimized for querying data.
- Snapshots are used on the write model side to optimally regenerate an aggregate state. Aggregate encapsulate your business logic and invariants (specific rules or conditions that must always hold true within a given system or domain) of your domain. So when you are writing/updating on the write model side, you will always validate against the current state of an aggregate to ensure data consistency at the aggregate level.

Conceptually both Projections and Snapshots are very different in terms of their usage. That said, Marten may internally use the constructs to provide these functionality but as an end user you can't interchangeably use these as you wish.

**Reply by jeremydmiller** [reply, marked as answer]
@federico-paolillo If you're putting the `Apply()` type methods directly on `Sensor`, just use:

`opts.Projections.SelfAggregate<Sensor>(lifecycle)`
